# Лабораторна робота №6. Функції

## Створення функції

Приклад функції з одним параметром і без ключового слова `return`:

In [1]:
def hi(name):
    print("Hi,", name)

hi("Greg")

Hi, Greg


Приклад функції з ключовим словом `return`:

In [2]:
def strange_function(n):
    if(n % 2 == 0):
        return True

print(strange_function(2))
print(strange_function(1))
    
    

True
None


### Завдання 1

Написати і протестувати фун
кцію, яка приймає один аргумент (рік) і повертає `True`, якщо рік є високосним, або `False` інакше.

Основу функції вже закладено у коді в редакторі.

__Примітка:__ для вас підготувлено короткий тестуючий код, який можна використовувати для перевірки своєї функції.

У коді використовуються два списки - один із тестовими даними, інший - з очікуваними результатами. Код повідомить Вам, якщо якісь із Ваших результатів неправильні.

In [3]:
def is_year_leap(year):
    if year < 1582:
        print("Not within the Gregorion calendar period")
        return False
    elif year % 4:
        return False
    elif year % 100:
        return True
    elif year % 400:
        return False
    else:
        return True


test_data = [1900, 2000, 2016, 1987]
test_results = [False, True, True, False]
for i in range(len(test_data)):
    yr = test_data[i]
    print(yr, "->", end="")
    result = is_year_leap(yr)
    if result == test_results[i]:
        print("OK")
    else:
        print("Failed")

1900 ->OK
2000 ->OK
2016 ->OK
1987 ->OK


### Завдання 2 

Написати та протестувати функцію, яка приймає два аргументи (рік і місяць) і повертає кількість днів для даної пари рік-місяць (у той час, як тільки лютий чутливий до значення `year`, Ваша функція має бути універсальною).

Початок функції готовий. Тепер переконайте функцію повернути `None`, якщо її аргументи не мають сенсу.

Звичайно, Ви можете (і повинні) використовувати раніше написану та протестовану функцію (див. попередні лаб. роб.). Це може бути дуже корисним. Ми рекомендуємо використовувати список із зазначенням довжини місяців. Ви можете створити його всередині функції – цей трюк значно скоротить код.

Ми підготували тестуючий код. Розвійте його, щоб увімкнути більше тестових випадків.

In [4]:
def is_year_leap(year):
    if year < 1582:
        print("Not within the Gregorion calendar period")
        return False
    elif year % 4:
        return False
    elif year % 100:
        return True
    elif year % 400:
        return False
    else:
        return True


def days_in_month(year, month):
    array_month = [31, None, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    if year < 1582 or month < 1 or month > 12:
        return None
    elif month == 2:
        if is_year_leap(year):
            return 29
        else:
            return 28
    else:
        other_months = array_month[month - 1]
        return other_months


test_years = [1900, 2000, 2016, 1987]
test_months = [2, 2, 1, 11]
test_results = [28, 29, 31, 30]
for i in range(len(test_years)):
    yr = test_years[i]
    mo = test_months[i]
    print(yr, mo, "->", end="")
    result = days_in_month(yr, mo)
    if result == test_results[i]:
        print("OK")
    else:
        print("Failed")


1900 2 ->OK
2000 2 ->OK
2016 1 ->OK
1987 11 ->OK


### Завдання 3

Написати та протестувати функцію, яка приймає три аргументи (рік, місяць та день місяця) та повертає відповідний день року або None, якщо якийсь із аргументів невірний.

Використовуйте раніше написані та протестовані функції. Додайте до коду свої власні тестові випадки.

In [5]:
def is_year_leap(year):
    if year < 1582:
        print("Not within the Gregorion calendar period")
        return False
    elif year % 4:
        return False
    elif year % 100:
        return True
    elif year % 400:
        return False
    else:
        return True


def days_in_month(year, month):
    array_month = [31, None, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    if year < 1582 or month < 1 or month > 12:
        return None
    elif month == 2:
        if is_year_leap(year):
            return 29
        else:
            return 28
    else:
        other_months = array_month[month - 1]
        return other_months


def day_of_year(year, month, day):
    if year < 1582 or month < 1 or month > 12 or day < 1 or day > days_in_month(year, month):
        return None
    else:
        month -= 1
        days = day
        for j in range(0, month):
            days += days_in_month(year, month)
            month -= 1
        return days


print(day_of_year(2000, 12, 31))


366


###  Завдання 4

Натуральне число є простим, якщо воно більше 1 і не має дільників, крім 1 і самого себе.

Важко? Зовсім ні. Наприклад, 8 не є простим числом, оскільки Ви можете розділити його на 2 та 4 (ми не можемо використовувати дільники, рівні 1 та 8, оскільки визначення забороняє це).

З іншого боку, 7 - просте число, оскільки ми можемо знайти йому підходящих дільників.


Ваше завдання - написати функцію, що перевіряє, чи є число простим чи ні.

Функція:

* називається `is_prime`;
* приймає один аргумент (значення для перевірки)
* повертає `True`, якщо аргумент є простим числом, і `False` інакше.  

__Підказка:__ спробуйте розділити аргумент на всі наступні значення (починаючи з 2) і перевірте залишок - якщо він дорівнює нулю, Ваше число не може бути простим; добре подумайте, коли вам слід зупинити процес.

Якщо вам потрібно знайти квадратний корінь із будь-якого значення, Ви можете використовувати оператор `**`. Пам'ятайте: квадратний корінь `x` дорівнює $x^{0,5}$.

Доповніть код у редакторі.

In [6]:
def is_prime(num):
    flag = True
    for x in range(2, num):
        if num % x == 0:
            flag = False
            break
    return flag


for i in range(1, 20):
    if is_prime(i + 1):
        print(i + 1, end=" ")
print()

2 3 5 7 11 13 17 19 


### Очікуваний вивод

`2 3 5 7 11 13 17 19`

### Завдання 5

Витрата палива автомобіля можна висловити по-різному. Наприклад, у Європі він відображається як кількість витраченого палива на 100 кілометрів шляху.

У США він відображається як кількість миль, пройдених автомобілем під час витрачання одного галону палива.

Ваше завдання - написати дві функції, що конвертують л/100км в миль на галон, і навпаки.

Функції:

* названі `liters_100km_to_miles_gallon` та `miles_gallon_to_liters_100km` відповідно;
* приймають один аргумент (значення, що відповідає їхнім іменам)  

Доповніть код у редакторі.

Запустіть свій код і перевірте, чи Ваш результат збігається з нашим.

Ось деяка інформація, яка може допомогти Вам:

* 1 американська мідя = 1609.344 метрів;
* 1 американський галон = 3.785411784 літрів.

In [7]:
def liters_100km_to_miles_gallon(liters):
    miles_of_km = 100 / (1609.344 / 1000)
    liters_100km = liters / 3.785411784
    miles_gallon = miles_of_km / liters_100km
    return miles_gallon


def miles_gallon_to_liters_100km(miles):
    mile_of_km = 1609.344 / 1000
    miles_gallon = (miles * mile_of_km) / 100
    liters_100km = 3.785411784 / miles_gallon
    return liters_100km


print(liters_100km_to_miles_gallon(3.9))
print(liters_100km_to_miles_gallon(7.5))
print(liters_100km_to_miles_gallon(10.))
print(miles_gallon_to_liters_100km(60.3))
print(miles_gallon_to_liters_100km(31.4))
print(miles_gallon_to_liters_100km(23.5))

60.31143162393162
31.36194444444444
23.52145833333333
3.9007393587617467
7.490910297239916
10.009131205673757


### Очікуваний вивод

`60.31143162393162`  
`31.36194444444444`  
`23.52145833333333`  
`3.9007393587617467`  
`7.490910297239916`  
`10.009131205673757`

### Приклад 1

Давайте напишемо функцію оцінки індексу маси тіла (ІМТ, BMI).
$$BMI = \frac{m}{h^2}$$
тут $m$ -- маса у _кг_, $h$ -- зріст у _м_.

In [16]:
def bmi(weight, height):
    return weight / height ** 2


print(bmi(52.5, 1.65))

19.283746556473833


### Ключове слово `none` 

Функція у попередньому прикладі виправдовує наші очікування, але вона дещо проста – вона припускає, що значення обох параметрів завжди мають сенс. Варто перевірити, чи заслуговують вони на довіру.

Давайте перевіримо їх обидва і повернемо `None`, якщо якесь з них виглядає підозріло.

In [19]:
def bmi(weight, height):
    if height < 1.0 or height > 2.5 or \
    weight < 20 or weight > 200:
        return None

    return weight / height ** 2


print(bmi(352.5, 1.65))

None


### Завдання 6 

Написати функцію, яка перевіряє, чи три сторони заданої довжини побудувати трикутник. Вона має повернути `True`, якщо сторони можуть утворити трикутник, і `False` в іншому випадку. У цьому випадку `is_a_triangle(a, b, c)` – гарна назва для такої функції.

In [10]:
def is_a_triangle(a, b, c):
    if a + b >= c and b + c >= a and c + a >= b:
        return True
    else:
        return False


a = float(input("введіть значення для сторони a: "))
b = float(input("введіть значення для сторони b: "))
c = float(input("введіть значення для сторони c: "))

print(is_a_triangle(a, b, c))

введіть значення для сторони a: 5
введіть значення для сторони b: 6
введіть значення для сторони c: 7
True


### Завдання 7 

Написати функцію `is_a_right_triangle(a, b, c)`, яка перевіряє, чи є трикутник зі сторонами `a`, `b`, `c` прямокутним. При цьому потрібно використати функцію `is_a_triangle(a, b, c)` з завдання 6. 

In [11]:
def is_a_triangle(a, b, c):
    if a + b >= c and b + c >= a and c + a >= b:
        return True
    else:
        return False


a = float(input("введіть значення для сторони a: "))
b = float(input("введіть значення для сторони b: "))
c = float(input("введіть значення для сторони c: "))

print(is_a_triangle(a, b, c))


def is_a_right_triangle(a, b, c):
    if (a ** 2) + (b ** 2) == (c ** 2) or (b ** 2) + (c ** 2) == (a ** 2) or (c ** 2) + (a ** 2) == (
            b ** 2) and is_a_triangle(a, b, c):
        return True
    else:
        return False


if is_a_right_triangle(a, b, c):
    print("прямокутний")
else:
    print("не прямокутний")

введіть значення для сторони a: 3
введіть значення для сторони b: 6
введіть значення для сторони c: 7
True
не прямокутний


### Контрольні запитання

1. Який вивод наступного фрагмента?  
![](image/lab6_1.jpg)
- помилка

2. Який вивод наступного фрагмента?  
![](image/lab6_2.jpg)
- 2 <br> 1

3. Який вивод наступного фрагмента?  
![](image/lab6_3.jpg)
- 2 <br> 3

4. Який вивод наступного фрагмента?  
![](image/lab6_4.jpg)
- 2 <br> 2